In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
train = pd.read_csv('NewFinal50.csv')

In [2]:
train

Unnamed: 0    parcelid  logerror  bathroomcnt  bedroomcnt  \
0                0  11016594.0  0.027600          2.0         3.0   
1                1  14366692.0 -0.168400          3.5         4.0   
2                2  12098116.0 -0.004000          2.0         3.0   
3                3  12643413.0  0.021800          2.0         3.0   
4                4  14432541.0 -0.005000          2.0         3.0   
5                5  11509835.0 -0.270500          2.0         3.0   
6                6  12286022.0  0.044000          2.0         3.0   
7                7  17177301.0  0.163800          2.0         3.0   
8                8  14739064.0 -0.003000          1.0         2.0   
9                9  14677559.0  0.084300          2.0         3.0   
10              10  10854446.0  0.382500          2.0         2.0   
11              11  11672170.0 -0.016100          4.0         5.0   
12              12  10949761.0  0.034400          2.0         3.0   
13              13  14665244.0  0.081600          2.0         3.0   
14              14  14148540.0 -0.045000          2.0         3.0   
15              15  14057104.0  0.002000          2.0         3.0   
16              16  17296797.0  0.044000          2.0         3.0   
17              17  12524288.0 -0.041900          1.0         1.0   
18              18  14091502.0 -0.002000          2.0         3.0   
19              19  10815909.0  0.010000          3.0         3.0   
20              20  12453853.0 -0.038700          1.0         2.0   
21              21  14253461.0  0.021800          2.0         3.0   
22              22  14001034.0  0.014900          2.0         3.0   
23              23  11458863.0  0.115100          2.0         3.0   
24              24  13957980.0 -0.169600          2.0         3.0   
25              25  12752047.0 -0.044000          4.0         3.0   
26              26  11188425.0 -0.074700          2.0         3.0   
27              27  14646051.0  0.008000          2.5         2.0   
28              28  10805063.0  0.001000          2.0         2.0   
29              29  12524725.0 -0.053400          2.0         3.0   
...            ...         ...       ...          ...         ...   
164819      164819  13089085.0  0.020421          3.0         4.0   
164820      164820  12906571.0  0.019484          2.0         3.0   
164821      164821  12651460.0  0.094750          1.0         2.0   
164822      164822  10962691.0 -0.064765          2.0         3.0   
164823      164823  11487400.0  0.046755          2.0         2.0   
164824      164824  12433997.0  0.018458          1.0         2.0   
164825      164825  10855745.0  0.002138          3.0         3.0   
164826      164826  12885530.0  0.003457          1.0         2.0   
164827      164827  11490983.0 -0.036067          1.0         2.0   
164828      164828  12401161.0 -0.232215          3.0         3.0   
164829      164829  11068972.0  0.015806          5.0         4.0   
164830      164830  11490089.0  0.068838          2.0         2.0   
164831      164831  12406123.0  0.000099          2.0         2.0   
164832      164832  12669704.0  0.047467          2.0         4.0   
164833      164833  12410050.0  0.000749          1.0         2.0   
164834      164834  11953526.0 -0.159471          0.0         0.0   
164835      164835  11978003.0 -0.062404          2.0         3.0   
164836      164836  11318911.0  0.013268          1.0         1.0   
164837      164837  10722237.0  0.052630          1.0         2.0   
164838      164838  12055795.0 -0.140881          1.0         2.0   
164839      164839  12892446.0  0.319408          3.0         4.0   
164840      164840  12666457.0  0.028733          3.0         3.0   
164841      164841  10858613.0 -0.075465          2.0         2.0   
164842      164842  10722691.0  0.081196          2.0         3.0   
164843      164843  12412492.0  0.001082          2.0         4.0   
164844      164844  10833991.0 -0.002245          3.0         3.0   
164845

In [3]:
train_ = train.drop(['logerror','logerror_labels','Unnamed: 0','parcelid'],1)
train_ 

bathroomcnt  bedroomcnt  calculatedbathnbr  \
0               2.0         3.0                2.0   
1               3.5         4.0                3.5   
2               2.0         3.0                2.0   
3               2.0         3.0                2.0   
4               2.0         3.0                2.0   
5               2.0         3.0                2.0   
6               2.0         3.0                2.0   
7               2.0         3.0                2.0   
8               1.0         2.0                1.0   
9               2.0         3.0                2.0   
10              2.0         2.0                2.0   
11              4.0         5.0                4.0   
12              2.0         3.0                2.0   
13              2.0         3.0                2.0   
14              2.0         3.0                2.0   
15              2.0         3.0                2.0   
16              2.0         3.0                2.0   
17              1.0         1.0                1.0   
18              2.0         3.0                2.0   
19              3.0         3.0                3.0   
20              1.0         2.0                1.0   
21              2.0         3.0                2.0   
22              2.0         3.0                2.0   
23              2.0         3.0                2.0   
24              2.0         3.0                2.0   
25              4.0         3.0                4.0   
26              2.0         3.0                2.0   
27              2.5         2.0                2.5   
28              2.0         2.0                2.0   
29              2.0         3.0                2.0   
...             ...         ...                ...   
164819          3.0         4.0                3.0   
164820          2.0         3.0                2.0   
164821          1.0         2.0                1.0   
164822          2.0         3.0                2.0   
164823          2.0         2.0                2.0   
164824          1.0         2.0                1.0   
164825          3.0         3.0                3.0   
164826          1.0         2.0                1.0   
164827          1.0         2.0                1.0   
164828          3.0         3.0                3.0   
164829          5.0         4.0                5.0   
164830          2.0         2.0                2.0   
164831          2.0         2.0                2.0   
164832          2.0         4.0                2.0   
164833          1.0         2.0                1.0   
164834          0.0         0.0                2.0   
164835          2.0         3.0                2.0   
164836          1.0         1.0                1.0   
164837          1.0         2.0                1.0   
164838          1.0         2.0                1.0   
164839          3.0         4.0                3.0   
164840          3.0         3.0                3.0   
164841          2.0         2.0                2.0   
164842          2.0         3.0                2.0   
164843          2.0         4.0                2.0   
164844          3.0         3.0                3.0   
164845          2.0         2.0                2.0   
164846          2.0         4.0                2.0   
164847          1.0         3.0                1.0   
164848          2.0         3.0                2.0   

        calculatedfinishedsquarefeet  finishedsquarefeet12    fips  \
0                             1544.0                1524.0  6037.0   
1                             2263.0                2263.0  6059.0   
2                             1544.0                1524.0  6037.0   
3                             1544.0                1524.0  6037.0   
4                             1544.0                1524.0  6037.0   
5                             1544.0                1524.0  6037.0   
6                             1544.0                1524.0  6037.0   
7                             1544.0                1524.0  6037.0   
8                            

In [4]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_predict
from sklearn import metrics

X = train_.values
Y = train["logerror"]

model = LinearRegression ()


F= train_.columns
f = len(F)

predicted0 = cross_val_predict(model, X, Y, cv=5)
score0 = metrics.mean_squared_error(Y, predicted0)
score = score0
print(score)

0.00932188872159


# Wrappers - backward selection
In backward selection, one starts with the set of all variables and progressively eliminates the least promising ones.

In [5]:
for i in range(f):
    # find which feature make the value of the evaluation function is optimal when remove this feature
    for j in range(len(F)):  
        train_1 = train_.drop([F[j]],1, inplace = False)
        X1 = train_1.values 
        predicted1 = cross_val_predict(model, X1, Y, cv=5)
        score1 = metrics.mean_squared_error(Y, predicted1)
        if score1 < score:
            score = score1
            k = j
    # remove this feature
    if score0 != score:         
        train_ = train_.drop([F[k]],1)
        F = F.drop(F[k],1)
        score0 = score
    # when we remove anyone feature, the value of the evaluation function cannot be better , we get the best feature
    else:
        print(score)
        break
print(F)              

0.00803071231006
Index(['calculatedbathnbr', 'finishedsquarefeet12', 'fips', 'fullbathcnt',
       'longitude', 'lotsizesquarefeet', 'propertycountylandusecode',
       'propertyzoningdesc', 'structuretaxvaluedollarcnt', 'taxvaluedollarcnt',
       'assessmentyear', 'landtaxvaluedollarcnt', 'taxamount', 'year0', '02',
       '03', '04', '05', '06', '07', '08'],
      dtype='object')


As we can see here, the best subset of features is 
0.00803071231006
['calculatedbathnbr', 'finishedsquarefeet12', 'fips', 'fullbathcnt',
       'longitude', 'lotsizesquarefeet', 'propertycountylandusecode',
       'propertyzoningdesc', 'structuretaxvaluedollarcnt', 'taxvaluedollarcnt',
       'assessmentyear', 'landtaxvaluedollarcnt', 'taxamount', 'year0', '02',
       '03', '04', '05', '06', '07', '08'].
